concent_api/core/models.py

```diff
@@ -28,6 +33,43 @@
 from .constants             import MESSAGE_TASK_ID_MAX_LENGTH
 
 
+class SubtaskWithTimingColumnsManager(Manager):
+    """Creates maximum_download time, subtask_verification_time and download_deadline columns
+    maximum_download_time = DOWNLOAD_LEADIN_TIME + ceil((result_package_size / MINIMUM_UPLOAD_RATE << 10))
+    subtask_verification_time = (4 * CONCENT_MESSAGING_TIME) + (3 * maximum_download_time) + (0.5 * (computation_deadline - task_to_compute_timestamp))
+    download_deadline = computation_deadline + maximum_download_time
+    """
+
+    def with_maximum_download_time(self):  # pylint: disable=no-self-use
+        bytes_per_sec = settings.MINIMUM_UPLOAD_RATE << 10
+        download_time = Func(F('result_package_size') / float(bytes_per_sec), function='CEIL')
+        return Value(settings.DOWNLOAD_LEADIN_TIME) + download_time
+
+    def with_subtask_verification_time(self):  # pylint: disable=no-self-use
+        computation_deadline = F('computation_deadline')
+        task_to_compute_timestamp = Func(Value('epoch'), F('task_to_compute__timestamp'), function='DATE_PART')
+        subtask_timeout = Func(Value('epoch'), computation_deadline, function='DATE_PART') - task_to_compute_timestamp
+        subtask_verification_time = (4 * settings.CONCENT_MESSAGING_TIME) + (3 * F('maximum_download_time')) + (
+                0.5 * subtask_timeout)
+        return ExpressionWrapper(subtask_verification_time, output_field=IntegerField())
+
+    def with_download_deadline(self):  # pylint: disable=no-self-use
+        computation_deadline = F('computation_deadline')
+        download_deadline = Func(Value('epoch'), computation_deadline, function='DATE_PART') + F(
+            'subtask_verification_time')
+        return ExpressionWrapper(download_deadline, output_field=IntegerField())
+
+    def with_timing_columns(self):
+        return super().get_queryset().annotate(
+            maximum_download_time=self.with_maximum_download_time(),
+            subtask_verification_time=self.with_subtask_verification_time(),
+            download_deadline=self.with_download_deadline(),
```

Comment:
It was not supposed to work like that. Each of these methods was meant to add a single column. Here you have one function that adds all three colunms.

Please make each function accept a queryset and return a new queryset with a single column added.

Reply:
It's not working when adding in separate colums.